### PR TITLE
Feature: Improvements to `automaticallyAdjustKeyboardInsets`

### PR DESCRIPTION
**Pull request by adamaveray**
## Summary:

This is a reopened version of #35224 by @isidoro98 which was closed without explanation, updated to resolve new merge conflicts and now includes an example in the RN-Tester app. Aside from that it is unchanged. Here is @isidoro98's description from their original PR:

This PR builds on top of #31402, which introduced the `automaticallyAdjustsScrollIndicatorInsets` functionality. It aims to fix one of RN's longstanding pain point regarding the keyboard.

The changes provide a better way of handling the `ScrollView` offset when a keyboard opens. Currently, when a keyboard opens we apply an **offset** to the `Scrollview` that matches the size of the keyboard. This approach is great if we are using an `InputAccessoryView` but if we have multiple `TextInputs` in a `ScrollView`; offsetting the content by the size of the keyboard doesn't yield the best user experience.

## Changelog:

[iOS] [Changed] - Scroll `ScrollView` text fields into view with `automaticallyAdjustsScrollIndicatorInsets`

## Test Plan:

The videos below compare the current and proposed behaviors for the following code:

```js
<ScrollView
  automaticallyAdjustKeyboardInsets
  keyboardDismissMode="interactive">
  {[...Array(10).keys()].map(item => (
    <CustomTextInput placeholder={item.toString()} key={item} />
  ))}
</ScrollView>
```

| Current behaviour | Proposal |
|-|-|
| https://user-images.githubusercontent.com/25139053/200194972-1ac5f1cd-2d61-4118-ad77-95c04d30c98d.mov | https://user-images.githubusercontent.com/25139053/200194990-53f28296-be11-4a47-be70-cec917d7deb1.mov |

As can be seen in the video, the **current behavior** applies an offset to the `ScrollView` content regardless of where the `TextInput` sits on the screen.

The proposal checks if the `TextInput` will be covered by the keyboard, and only then applies an offset. The offset applied is not the full size of the keyboard but instead only the required amount so that the `TextInput` is a **specific** distance above the top of the keyboard (customizable using the new `bottomKeyboardOffset` prop). This achieves a less "jumpy" experience for the user.

The proposal doesn't change the behavior of the `ScrollView` offset when an `InputAccessory` view is used, since it checks if the `TextField` that triggered the keyboard is a **descendant** of the `ScrollView` or not.

## Why not use other existing solutions?

RN ecosystem offers other alternatives for dealing with a keyboard inside a ScrollView, such as a `KeyboardAvoidingView` or using third party libraries like `react-native-keyboard-aware-scroll-view`. But as shown in the recordings below, these solutions don't provide the smoothness or behavior that can be achieved with `automaticallyAdjustsScrollIndicatorInsets`.

| KeyboardAvoidingView | rn-keyboard-aware-scroll-view |
|-|-|
| https://user-images.githubusercontent.com/25139053/200195145-de742f0a-6913-4099-83c4-7693448a8933.mov | https://user-images.githubusercontent.com/25139053/200195151-80745533-16b5-4aa0-b6cd-d01041dbd001.mov |

As shown in the videos, the `TextInput` is hidden by the keyboard for a split second before becoming visible.

Code for the videos above:

```js
// KeyboardAvoidingView
<KeyboardAvoidingView
  style={{flex: 1, flexDirection: 'column', justifyContent: 'center'}}
  behavior="padding"
  enabled>
  <ScrollView>
    {[...Array(10).keys()].map(item => (
      <CustomTextInput placeholder={item.toString()} key={item} />
    ))}
  </ScrollView>
</KeyboardAvoidingView>
```

 ```js
// rn-keyboard-aware-scroll-view
<KeyboardAwareScrollView>
  {[...Array(10).keys()].map(item => (
    <CustomTextInput placeholder={item.toString()} key={item} />
  ))}
</KeyboardAwareScrollView>
```

### CODE REVIEW

File: packages/react-native/React/Views/UIResponder+FirstResponder.m [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+#import "UIResponder+FirstResponder.h"
+
+static __weak id currentFirstResponder;
+
+@implementation UIResponder (FirstResponder)
+
++(id)currentFirstResponder {
+    currentFirstResponder = nil;
+    [[UIApplication sharedApplication] sendAction:@selector(findFirstResponder:) to:nil from:nil forEvent:nil];
+    return currentFirstResponder;
```
Comment (javache) [inline packages/react-native/React/Views/UIResponder+FirstResponder.m:17]: This is a private API, which will get your app rejected from the App Store.

**Comment by javache** [comment]
Why can't `bottomKeyboardOffset` be implemented using a `marginBottom` on the TextInput component?

**Review by javache — CHANGES_REQUESTED** [review]

File: packages/react-native/React/Views/ScrollView/RCTScrollView.m [hunk 2]
```diff
@@ -324,11 +326,26 @@ - (void)_keyboardWillChangeFrame:(NSNotification *)notification
   }
 
   CGPoint newContentOffset = _scrollView.contentOffset;
-  CGFloat contentDiff = endFrame.origin.y - beginFrame.origin.y;
-  if (self.inverted) {
-    newContentOffset.y += contentDiff;
+  UIResponder *firstResponder = [UIResponder currentFirstResponder];
+  if ([firstResponder isKindOfClass: [UITextField class]] && [(UITextField *) firstResponder isDescendantOfView:_scrollView]) {
+    UITextField *textField = [UIResponder currentFirstResponder];
```
Comment (javache) [inline packages/react-native/React/Views/ScrollView/RCTScrollView.m:331]: You could structure this using `sendAction` perhaps.

```suggestion

  UIResponder *firstResponder = [UIResponder currentFirstResponder];
  if ([firstResponder isKindOfClass: [UITextField class]] && [(UITextField *) firstResponder isDescendantOfView:_scrollView]) {
    UITextField *textField = [UIResponder currentFirstResponder];
```

**Comment by adamaveray** [comment]
@javache Thanks for reviewing and for your feedback. This PR was initially put together by another contributor so I wasn't aware of the private API usage. I've overhauled how it's all implemented now to avoid that and to remove the niche `bottomKeyboardOffset` value. It now replicates native iOS focus behaviour as closely as I could get it, which can be tested in the updated RN Tester ScrollViewKeyboardInsets screen. Please let me know if you have any feedback on the changes!

File: packages/react-native/Libraries/Text/TextInput/RCTBaseTextInputView.m [hunk 3]
```diff
@@ -19,6 +20,8 @@
 #import <React/RCTTextAttributes.h>
 #import <React/RCTTextSelection.h>
 
+#define EMPTY_TEXT_KEYBOARD_BOTTOM_OFFSET 15 // Native iOS empty text field bottom keyboard offset amount
```
Comment (javache) [inline packages/react-native/Libraries/Text/TextInput/RCTBaseTextInputView.m:23]: Prefer `static const CGFloat *kEmptyKeyboardBottomOffset` over #define (which doesn't have any any type info)

File: packages/react-native/Libraries/Text/TextInput/RCTBaseTextInputView.m [hunk 4]
```diff
@@ -27,6 +30,31 @@ @implementation RCTBaseTextInputView {
...
+
+  // Focus on text selection frame
+  CGRect focusRect = CGRectMake(
+    selection.rect.origin.x,
+    selection.rect.origin.y,
+    selection.rect.size.width,
+    selection.rect.size.height + (selectedTextRange.empty ? EMPTY_TEXT_KEYBOARD_BOTTOM_OFFSET : 0)
+  );
+  scrollView.firstResponderFocus = [self convertRect:focusRect toView:nil];
+}
```
Comment (javache) [inline packages/react-native/Libraries/Text/TextInput/RCTBaseTextInputView.m:56]: Avoid early return if both branches are similar
```suggestion
  if (selection == nil) {
    // No active selection or caret - fallback to entire input frame
    scrollView.firstResponderFocus = [self convertRect:self.frame toView:nil];
  } else {
    // Focus on text selection frame
    CGRect focusRect = CGRectMake(
      selection.rect.origin.x,
      selection.rect.origin.y,
      selection.rect.size.width,
      selection.rect.size.height + (selectedTextRange.empty ? EMPTY_TEXT_KEYBOARD_BOTTOM_OFFSET : 0)
    );
    scrollView.firstResponderFocus = [self convertRect:focusRect toView:nil];
   }
}
```

File: packages/react-native/Libraries/Text/TextInput/RCTBaseTextInputView.m [hunk 5]
```diff
@@ -27,6 +30,31 @@ @implementation RCTBaseTextInputView {
...
+  if (![self isDescendantOfView:scrollView]) {
+    // View is outside scroll view
+    return;
+  }
+
+  UITextRange *selectedTextRange = self.backedTextInputView.selectedTextRange;
+  UITextSelectionRect *selection = [self.backedTextInputView selectionRectsForRange:selectedTextRange].firstObject;
+  if (selection == nil) {
+    // No active selection or caret - fallback to entire input frame
+    scrollView.firstResponderFocus = [self convertRect:self.frame toView:nil];
```
Comment (javache) [inline packages/react-native/Libraries/Text/TextInput/RCTBaseTextInputView.m:44]: `convertRect` is in the receiver's coordinate system, so this needs to be self.bounds

File: packages/react-native/React/Views/ScrollView/RCTScrollView.m [hunk 6]
```diff
@@ -324,7 +325,20 @@ - (void)_keyboardWillChangeFrame:(NSNotification *)notification
   }
 
   CGPoint newContentOffset = _scrollView.contentOffset;
-  CGFloat contentDiff = endFrame.origin.y - beginFrame.origin.y;
+  self.firstResponderFocus = CGRectNull;
+
+  CGFloat contentDiff = 0;
+  if ([[UIApplication sharedApplication] sendAction:@selector(reactUpdateResponderOffsetForScrollView:) to:nil from:self forEvent:nil]) {
+    // Inner text field focused
+    CGFloat focusEnd = self.firstResponderFocus.origin.y + self.firstResponderFocus.size.height;
```
Comment (javache) [inline packages/react-native/React/Views/ScrollView/RCTScrollView.m:333]:
```suggestion
    CGFloat focusEnd = CGRectGetMaxY(self.firstResponderFocus);
```

File: packages/react-native/React/Views/ScrollView/RCTScrollView.m [hunk 7]
```diff
@@ -324,7 +325,20 @@ - (void)_keyboardWillChangeFrame:(NSNotification *)notification
...
+    if (focusEnd > endFrame.origin.y && focusEnd != INFINITY) {
```
Comment (javache) [inline packages/react-native/React/Views/ScrollView/RCTScrollView.m:334]: When can you ever get `INFINITY` here?

File: packages/react-native/React/Views/ScrollView/RCTScrollView.m (hunk 6 above)
Comment (javache) [reply packages/react-native/React/Views/ScrollView/RCTScrollView.m:333]: This code currently assumes the scrollview is vertical, it should really work when content is laid out horizontally too.

**Review by javache — COMMENTED** [review]
@philIip Any concerns with this? You mentioned some changes coming to keyboard layout events with iOS17.

File: packages/react-native/React/Views/ScrollView/RCTScrollView.m (hunk 6 above)
Comment (adamaveray) [reply packages/react-native/React/Views/ScrollView/RCTScrollView.m:333]: @javache This `_keyboardWillChangeFrame:` method has an existing `if ([self isHorizontal:_scrollView]) { return; }` check at the top so this code only applies to vertical scroll views. Fortunately when the scroll view is horizontal the default behaviour correctly scrolls partially-obscured text fields into view when focused.

File: packages/react-native/React/Views/ScrollView/RCTScrollView.m (hunk 7 above)
Comment (adamaveray) [reply packages/react-native/React/Views/ScrollView/RCTScrollView.m:334]: @javache It happens when the focussed text view is not a child of the scroll view. I've updated the code to make that clearer.

**Comment by adamaveray** [comment]
@javache Thanks for the additional review; I've made the requested changes and also done some additional testing & updates to more closely replicate the native bottom keyboard offset amount (which I initially thought was due to an empty selection but have since found it's actually single line aka UITextField vs multiline aka UITextView). Please let me know if you have any more feedback.

**Comment by adamaveray** [comment]
@javache Just checking in to see if there are any other issues you can see with this PR that are standing in the way of it potentially being merged?

File: packages/react-native/Libraries/Text/TextInput/RCTBaseTextInputView.m [hunk 8]
```diff
@@ -27,6 +31,31 @@ @implementation RCTBaseTextInputView {
...
+    scrollView.firstResponderFocus = [self convertRect:self.bounds toView:nil];
+  } else {
+    // Focus on text selection frame
+    BOOL isMultiline = [self.backedTextInputView isKindOfClass:[UITextView class]];
+    CGRect focusRect = CGRectMake(
+      selection.rect.origin.x,
+      selection.rect.origin.y,
+      selection.rect.size.width,
+      selection.rect.size.height + (isMultiline ? 0 : kSingleLineKeyboardBottomOffset)
+    );
```
Comment (javache) [inline packages/react-native/Libraries/Text/TextInput/RCTBaseTextInputView.m:54]:
```suggestion
    CGRect focusRect = selection.rect;
    focusReact.size.height += (isMultiline ? 0 : kSingleLineKeyboardBottomOffset)
```

File: packages/react-native/React/Views/ScrollView/RCTScrollView.h [hunk 9]
```diff
@@ -48,6 +48,7 @@
 @property (nonatomic, assign) BOOL snapToEnd;
 @property (nonatomic, copy) NSString *snapToAlignment;
 @property (nonatomic, assign) BOOL inverted;
+@property (nonatomic, assign) CGRect firstResponderFocus;
```
Comment (javache) [inline packages/react-native/React/Views/ScrollView/RCTScrollView.h:51]: We should document what this rect should be relative to, right now it seems to be to be the window, but should it not be the scrollview itself?

Can you test with scrollviews that do not to take up the entire screen size?

**Comment by javache** [comment]
Thanks for adding the example (and sorry for the delay)! This mostly looks good to me, but would like to clarify what the coordinates used are relative to.

File: packages/react-native/Libraries/Text/TextInput/RCTBaseTextInputView.m (hunk 8 above)
Comment (adamaveray) [reply packages/react-native/Libraries/Text/TextInput/RCTBaseTextInputView.m]: Thanks for the suggestion! I've just incorporated that and reworked that section more broadly to reduce duplication and try and make it all a bit more obvious what going on.

**Comment by adamaveray** [comment]
Thanks for following up! I've implemented those requests & suggestions, let me know if you have any other feedback 🙌

**Comment by javache** [comment]
This looks good to merge now, but there are conflicts with main in `RCTBaseTextInputView.m`. Could you resolve those?

**Comment by adamaveray** [comment]
@javache The merge conflicts should be resolved now. Let me know if anything else needs updating!

**Comment by javache** [comment]
Thanks for contributing! It'd be great if you could make this feature compatible with the new architecture too.

**Pull request closed without merging**